File: examples/solve-recaptchas/frameworks/playwright/java/CaptchaExample.java
// Connects to Browserless with solveCaptchas=true and waits for the
// Browserless.captchaAutoSolved CDP event before submitting the form.
//
// Reuse the existing page from the default context so Browserless CDP events
// (including captchaAutoSolved) are visible on this page object. Register the
// CDP listener before navigation so the event isn't missed if the CAPTCHA solves
// immediately after the page loads.
//
// Add to pom.xml:
//   <dependency>
//     <groupId>com.microsoft.playwright</groupId>
//     <artifactId>playwright</artifactId>
//     <version>1.44.0</version>
//   </dependency>
//
// Compile: mvn compile
// Run:     mvn exec:java

import com.microsoft.playwright.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

public class CaptchaExample {
    public static void main(String[] args) throws Exception {
        String wsEndpoint =
            "wss://production-sfo.browserless.io/stealth"
            + "?token=YOUR_API_TOKEN_HERE"
            + "&proxy=residential&proxyCountry=us"
            + "&solveCaptchas=true&timeout=300000";

        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(wsEndpoint);
            try {
                // Reuse the existing page from the default context so Browserless CDP
                // events (including captchaAutoSolved) are visible on this page object.
                BrowserContext context = browser.contexts().get(0);
                Page page = context.pages().get(0);
                CDPSession cdp = page.context().newCDPSession(page);

                // Register before navigation so the event isn't missed if the CAPTCHA
                // solves immediately after the page loads.
                CompletableFuture<Void> captchaSolved = new CompletableFuture<>();
                cdp.on("Browserless.captchaAutoSolved", event -> captchaSolved.complete(null));

                page.navigate(
                    "https://www.google.com/recaptcha/api2/demo",
                    new Page.NavigateOptions().setWaitUntil(WaitUntilState.NETWORKIDLE)
                );

                // Await the CDP event rather than a fixed timeout — fires when Browserless
                // finishes solving the CAPTCHA, not after an arbitrary number of seconds.
                captchaSolved.get(30, TimeUnit.SECONDS);

                page.click("#recaptcha-demo-submit");
                page.waitForLoadState(LoadState.NETWORKIDLE);
                System.out.println("Done. Final URL: " + page.url());
            } finally {
                // Always close to release the session even on error.
                browser.close();
            }
        }
    }
}

File: examples/solve-recaptchas/frameworks/playwright/csharp/CaptchaExample.cs
// Connects to Browserless with solveCaptchas=true and waits for the
// Browserless.captchaAutoSolved CDP event before submitting the form.
//
// Reuse the existing page from the default context so Browserless CDP events
// (including captchaAutoSolved) are visible on this page object. Register the
// CDP listener before navigation so the event isn't missed if the CAPTCHA solves
// immediately after the page loads.
//
// Install: dotnet add package Microsoft.Playwright
// Run:     dotnet run

using Microsoft.Playwright;
using System;
using System.Threading.Tasks;

const string WsEndpoint =
    "wss://production-sfo.browserless.io/stealth"
    + "?token=YOUR_API_TOKEN_HERE"
    + "&proxy=residential&proxyCountry=us"
    + "&solveCaptchas=true&timeout=300000";

using var playwright = await Playwright.CreateAsync();
var browser = await playwright.Chromium.ConnectOverCDPAsync(WsEndpoint);

try
{
    // Reuse the existing page from the default context so Browserless CDP
    // events (including captchaAutoSolved) are visible on this page object.
    var context = browser.Contexts[0];
    var page = context.Pages[0];
    var cdp = await page.Context.NewCDPSessionAsync(page);

    // Register before navigation so the event isn't missed if the CAPTCHA
    // solves immediately after the page loads.
    var captchaSolved = new TaskCompletionSource<bool>();
    cdp.On("Browserless.captchaAutoSolved", e => captchaSolved.TrySetResult(true));

    await page.GotoAsync(
        "https://www.google.com/recaptcha/api2/demo",
        new PageGotoOptions { WaitUntil = WaitUntilState.NetworkIdle }
    );

    // Await the CDP event rather than a fixed timeout — fires when Browserless
    // finishes solving the CAPTCHA, not after an arbitrary number of seconds.
    // Compare completed task to timeoutTask so the exception is observed and thrown.
    var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
    var completed = await Task.WhenAny(captchaSolved.Task, timeoutTask);
    if (completed == timeoutTask)
        throw new TimeoutException("CAPTCHA timeout");

    await page.ClickAsync("#recaptcha-demo-submit");
    await page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    Console.WriteLine($"Done. Final URL: {page.Url}");
}
finally
{
    // Always close to release the session even on error.
    await browser.CloseAsync();
}
